rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,348 @@
+# User Uploads
+
+## For the cellxgene Data Portal
+
+**Authors:** [Trent Smith](mailto:trent.smith@chanzuckerberg.com)
+
+**Approvers:** [Friendly Tech Person](mailto:some.nice.person@chanzuckerberg.com), [Friendly PM Person](mailto:some.nice.person@chanzuckerberg.com), [Friendly Tech Person2](mailto:some.nice.person@chanzuckerberg.com), [Girish Patangay](mailto:girish.patangay@chanzuckerberg.com), [Sara Gerber](mailto:sara.gerber@chanzuckerberg.com)
+
+## tl;dr
+
+A user can upload files of any size to Data Portal using either a shared link from their cloud provider, or directly from their browser.
+
+## Glossary
+
+- CSP(Cloud Storage Provider) - Cloud storage services such as Dropbox or Google Drive.
+- Shared Link - A link from a CSP that give the bearer of that link access to private data.
+- DP(Data Portal) - The cellxgene Data Portal.
+- Small File - a file < 5 GB in size.
+- Large File - a file > 5 GB in size.
+
+## Problem Statement | Background
+
+One of the primary goals of the DP is to allow users to upload their own data sets. These data sets can vary in
+size from a few MB to several GB. User may want to upload from their local machine or from their CSP. For
+users that want to upload from their local machine, they must be able to resume an upload in the event their connection
+is interrupted. For users that want to upload data sets from their CSP, they are able to provide a shared link from
+their CSP that we can used to upload directly to our storage.
+
+## Product Requirements
+
+1. A user can upload a file to DP using a shared link from their CSP.
+1. A user can upload a file from their local machine using the DP browser.
+1. A user can resume an upload with out completely restarting.
+1. The user must be authenticated before uploading.
```

Comment:
The publisher must also agree to portal policies as outlined in [Publishers must complete attestation before submitting private datasets to the portal](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/22). The UX will enforce this behavior  but it will need to be included in the related REST API implementation.

Reply:
added

Reply:
Can you elaborate? If attestation is required, the user must upload the attestation file first before uploading datasets or any other file. The attestation flag indicating the attestation requirements have been met, live in the Data Portal DB. Therefore the backend API would need to block the user from uploading if the flag is not set.

Reply:
I will add in attestation to the REST API

Reply:
We need to review the proposed UX flow with Shawn, but the intent is to include a link to the appropriate terms and conditions for review during collection creation. The user must accept before datasets can be uploaded to the collection. It does not seem efficient to upload multiple copies of the same agreement to the database or duplicate the current Google Form experience.

Reply:
@Bento007 Maybe a clearer way to word this is "A user who has not yet agreed to the Portal policies must not be able to upload their file." The negative case is easier for me to understand because this upload service shouldn't be dealing with uploading attestation files but rather checking whether the user has done its attestation homework already and therefore can proceed.